Don't switch streams when running Caffe2 ops from c10.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55121 Don't switch streams when running Caffe2 ops from c10.**

This is done by allow -1 as a stream ID, meaning "don't change
the stream", in SwitchToDevice

Fixes #54830

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27527544](https://our.internmc.facebook.com/intern/diff/D27527544)

ezyang committed:
Don't switch streams when running Caffe2 ops from c10.

This is done by allow -1 as a stream ID, meaning "don't change
the stream", in SwitchToDevice

Fixes #54830

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
cc @jwestell, see if this fixes your problem

jwestell commented on caffe2/core/export_caffe2_op_to_c10.h:
```diff
@@ -29,7 +29,7 @@ inline c10::List<at::Tensor> _call_caffe2_op(
     c10::List<at::Tensor>&& outputs) {
   Caffe2Operator op(schema, std::move(inputs), std::move(outputs));
-  op.Run();
+  op.Run(-1);
```
@ezyang, I believe it is the line above this (constructor of `caffe2::Operator` [[operator.h#L814](https://github.com/pytorch/pytorch/blob/7824d8277aebc20ac8c21c5f4b52a29182c24816/caffe2/core/operator.h#L814)] that does the initial switching to device, and thus the additional stream is already created and switched to here.

ezyang replied:
Good catch.

ezyang committed:
Update on "Don't switch streams when running Caffe2 ops from c10."


This is done by allow -1 as a stream ID, meaning "don't change
the stream", in SwitchToDevice

Fixes #54830

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

jwestell commented:
The patch works great, @ezyang. My application stays on the defualt stream after using the caffe2 op. Thanks for your help.
FWIW, I tested this patch backported to the v1.7.0 tag.

ezyang committed:
Update on "Don't switch streams when running Caffe2 ops from c10."


This is done by allow -1 as a stream ID, meaning "don't change
the stream", in SwitchToDevice

Fixes #54830

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27527544](https://our.internmc.facebook.com/intern/diff/D27527544)

[ghstack-poisoned]
